### PR TITLE
Make runs-on support variable expression

### PR DESCRIPTION
**Pull request by sillyguodong**
As title.
Close issue: https://gitea.com/gitea/act_runner/issues/445
Follow: https://gitea.com/gitea/act/pulls/91

Move `getSecretsOfTask` and `getVariablesOfTask` under `models` because of circular dependency issues.

### CODE REVIEW

File: modules/actions/utils.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+// SPDX-License-Identifier: MIT
+
+package actions
+
+import (
+	"context"
+
+	actions_model "code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/models/db"
+	secret_model "code.gitea.io/gitea/models/secret"
```
Comment (KN4CK3R) [inline modules/actions/utils.go]: `modules` don't access the database.

**Review by KN4CK3R — DISMISSED** [review]

File: modules/actions/utils.go (hunk 1 above)
Comment (sillyguodong) [reply modules/actions/utils.go]: move to `service` package [14f2606](https://github.com/go-gitea/gitea/pull/29468/commits/14f2606023baf839ccd2a2bfc122291df2251ff0)

File: services/actions/utils.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	actions_model "code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/models/db"
+	secret_model "code.gitea.io/gitea/models/secret"
+	actions_module "code.gitea.io/gitea/modules/actions"
+	"code.gitea.io/gitea/modules/log"
+	secret_module "code.gitea.io/gitea/modules/secret"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+func GetSecretsOfTask(ctx context.Context, task *actions_model.ActionTask) map[string]string {
```
Comment (lunny) [inline services/actions/utils.go]: How about move this function to `secret.go`?

File: services/actions/utils.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			// go on
+		} else {
+			secrets[secret.Name] = v
+		}
+	}
+
+	return secrets
+}
+
+func GetVariablesOfRun(ctx context.Context, run *actions_model.ActionRun) map[string]string {
```
Comment (lunny) [inline services/actions/utils.go]: How about move this function to `variables.go`?

File: models/actions/variable.go [hunk 4]
```diff
@@ -82,3 +83,32 @@ func UpdateVariable(ctx context.Context, variable *ActionVariable) (bool, error)
...
 	return count != 0, err
 }
+
+func GetVariablesOfRun(ctx context.Context, run *ActionRun) map[string]string {
+	variables := map[string]string{}
+
+	// Global
+	globalVariables, err := db.Find[ActionVariable](ctx, FindVariablesOpts{})
+	if err != nil {
+		log.Error("find global variables: %v", err)
```
Comment (lunny) [inline models/actions/variable.go:93]: Why not return error?

File: models/secret/secret.go [hunk 5]
```diff
@@ -112,3 +115,39 @@ func UpdateSecret(ctx context.Context, secretID int64, data string) error {
...
+	if task.Job.Run.IsForkPullRequest && task.Job.Run.TriggerEvent != actions_module.GithubEventPullRequestTarget {
+		// ignore secrets for fork pull request, except GITHUB_TOKEN and GITEA_TOKEN which are automatically generated.
+		// for the tasks triggered by pull_request_target event, they could access the secrets because they will run in the context of the base branch
+		// see the documentation: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target
+		return secrets
+	}
+
+	ownerSecrets, err := db.Find[Secret](ctx, FindSecretsOptions{OwnerID: task.Job.Run.Repo.OwnerID})
+	if err != nil {
+		log.Error("find secrets of owner %v: %v", task.Job.Run.Repo.OwnerID, err)
```
Comment (lunny) [inline models/secret/secret.go:134]: The same as before.

File: models/actions/variable.go (hunk 4 above)
Comment (sillyguodong) [reply models/actions/variable.go:93]: done [eba48db](https://github.com/go-gitea/gitea/pull/29468/commits/eba48db10fb98482b1089f386a70f40a73b1889b)

File: models/actions/variable.go [hunk 6]
```diff
@@ -82,3 +83,34 @@ func UpdateVariable(ctx context.Context, variable *ActionVariable) (bool, error)
...
+	ownerVariables, err := db.Find[ActionVariable](ctx, FindVariablesOpts{OwnerID: run.Repo.OwnerID})
+	if err != nil {
+		log.Error("find variables of org: %d, error: %v", run.Repo.OwnerID, err)
+		return nil, err
+	}
+
+	// Repo level
+	repoVariables, err := db.Find[ActionVariable](ctx, FindVariablesOpts{RepoID: run.RepoID})
+	if err != nil {
+		log.Error("find variables of repo: %d, error: %v", run.RepoID, err)
```
Comment (lunny) [inline models/actions/variable.go:107]: Should this also return err?

**Review by lunny — APPROVED** [review]

**Comment by sillyguodong** [comment]
@KN4CK3R I have addressed the requested changes in the PR. Could you please re-review it? Thank you!

**Review by Zettat123 — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**